### ISSUE TITLE
Rails 7.1 Create model with has_many with lambda incorrect default

### ISSUE THREAD

**Issue by nikita-railsware** [rails/rails#50645]
### Steps to reproduce

In Rails 7.1, when I create new model with lambda in `has_many` association without passing value to the parameter, it puts serialized value of array that passed in `where` in a lambda.
Rails 7.0 just put `nil` in this case.

```ruby

# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", "7.1.2"
  # If you want to test against edge Rails replace the previous line with this:
  # gem "rails", github: "rails/rails", branch: "main"

  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end

  create_table :users, force: true do |t|
  end

  create_table :post_accesses, force: true do |t|
    t.string :access_type
    t.references :post
    t.references :user
  end
end

class PostAccess < ActiveRecord::Base
  ROLES = %w[viewer editor owner]
  STAFF_ROLES = %w[editor owner]

  belongs_to :user
  belongs_to :post
end

class Post < ActiveRecord::Base
  has_many :post_accesses
  has_many :staff_post_accesses,
           -> { where(access_type: PostAccess::STAFF_ROLES) },
           class_name: 'PostAccess'

  has_many :staff_users, through: :staff_post_accesses, class_name: 'User', source: :user
end

class User < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_association_stuff
    user = User.create!
    post = Post.create!(staff_users: [user])

    assert_equal nil, post.post_accesses.first.access_type
  end
end
```

### Expected behavior

It should put `nil`, or one of the possible value

### Actual behavior

It put serialized array:
'["editor", "owner"]'

### System configuration
**Rails version**: 7.1.2

**Ruby version**: 3.2.2

**Comment by bparanj** [comment]
The test is wrong. Here is the correct usage.

```ruby
class BugTest < Minitest::Test
  def test_association_stuff
    user = User.create!
    access_type = 'editor' # Choose appropriate access_type
    post_access = PostAccess.create!(user:, access_type:)
    post = Post.create!(post_accesses: [post_access])

    assert_equal access_type, post.post_accesses.first.access_type
  end
end
```

**Comment by fatkodima** [comment]
This started to fail after https://github.com/rails/rails/pull/47485 (cc @nvasilevski)

**Comment by nvasilevski** [comment]
This is an interesting one!

>Rails 7.0 just put nil in this case.

So I think that putting `nil` and actually creating a `staff_post_accesses` record with an empty `access_type` was incorrect. In a bit more simplified setup like:
```ruby
  has_many :staff_post_accesses, -> { where(access_type: 'owner') }, class_name: 'PostAccess'
```

Rails application would expect `staff_post_accesses` join records to be created with `access_type` populated as `owner` so I don't think we can revert the PR as a fix. However this is trickier when the querying parameter is an Array as it leads to what we see here.
I'm also curious, @nikita-railsware, would you mind elaborating how your application eventually sets the `access_type` column on the record since I assume technically an empty `access_type` value is invalid and you would want at least some value is required.  

I haven't tried it just yet but as a fix we could consider excluding attributes where value is an Array from creation here
https://github.com/Shopify/rails/blob/2801ea64c36852a1e07c9fcfa3785d56007d6052/activerecord/lib/active_record/associations/has_many_through_association.rb#L78

Though it feels a little hacky so what I'm leaning towards is actually changing the application a little bit. I think this may actually improve the handling and help setting the `access_type` properly:

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", "7.1.2"
  # If you want to test against edge Rails replace the previous line with this:
  # gem "rails", github: "rails/rails", branch: "main"

  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end

  create_table :users, force: true do |t|
  end

  create_table :post_accesses, force: true do |t|
    t.string :access_type
    t.references :post
    t.references :user
  end
end

class PostAccess < ActiveRecord::Base
  ROLES = %w[viewer editor owner]
  STAFF_ROLES = %w[editor owner]

  belongs_to :user
  belongs_to :post
end

class Post < ActiveRecord::Base
  has_many :post_accesses
  # this is purely for querying
  has_many :staff_post_accesses,
           -> { where(access_type: PostAccess::STAFF_ROLES) },
           class_name: 'PostAccess'

  # this is purely for querying
  has_many :staff_users, through: :staff_post_accesses, class_name: 'User', source: :user

  # used to create staff users with the `editor` access type, can be used for querying
  has_many :editor_staff_post_accesses, -> { where(access_type: 'editor') }, class_name: 'PostAccess'
  has_many :editor_staff_users, through: :editor_staff_post_accesses, class_name: 'User', source: :user


  # used to create staff users with the `owner` access type, can be used for querying
  has_many :owner_staff_post_accesses, -> { where(access_type: 'owner') }, class_name: 'PostAccess'
  has_many :owner_staff_users, through: :owner_staff_post_accesses, class_name: 'User', source: :user
  
end

class User < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_association_stuff
    user = User.create!
    post1 = Post.create!(editor_staff_users: [user])
    post2 = Post.create!(owner_staff_users: [user])

    assert_equal 'editor', post1.post_accesses.first.access_type
    assert_equal 'owner', post2.post_accesses.first.access_type
  end
end
```

I think that would be a nicer approach even though it's a little bit verbose and it heavily relies on the fact that eventually we do want the `access_type` to be populated somehow. 
But I think the same principle can be applied even if we don't want `access_type` to be populated. We could also define a separate association purely for creation purposes which won't have custom lambda.

Basically the conventional association:
```ruby
  has_many :post_accesses
  has_many :users, through: :post_accesses
```

And I think this is also a reasonable solution since technically it's a bit misleading to use `staff_users: [user]` association to create a user which won't have any indication of being a staff member (having `access_type` as `nil`)

**Comment by mattr** [comment]
Agree with @nvasilevski's assessment. This looks like it was possibly incorrect behaviour in prior versions of Rails. The alternative would be to create an association for each entry of the array, which could lead to unexpected behaviour. Using explicit associations as per the proposed solution makes it clearer what the associations are intended to do, and that the actual behaviour should be considered the expected behaviour.

**Comment by merouaneamqor** [comment]
Consider refactoring by creating distinct associations for each role – something like `editor_staff_post_accesses `and `owner_staff_post_accesses`. 
Each of these would have its own lambda clearly defining a singular role. This method trades a bit of verbosity for precision, effectively eliminating any potential confusion about how `access_type `gets determined. 
It’s like labeling your food in a shared fridge – clear, straightforward, and avoids any unwanted surprises.
